splinter/driver/webdriver/__init__.py

```diff
@@ -599,20 +599,35 @@ def check(self, name):
     def uncheck(self, name):
         self.find_by_name(name).first.uncheck()
 
-    def screenshot(self, name="", suffix=".png", full=False):
+    def screenshot(self, filename, full=False, waiting_time=0):
```

Comment:
It's still possible to keep the same signature like this:

```python
def screenshot(self, name="", suffix=".png", full=False, waiting_time=0):
        if full:
            self.full_screen()
            # trigger the `scroll` event to ensure lazy-load images can be rendered.
            self.execute_script("window.dispatchEvent(new Event('scroll'))") 
        if waiting_time > 0:
            time.sleep(waiting_time)

        if not name:
             (fd, name) = tempfile.mkstemp(prefix=name, suffix=suffix)
            os.close(fd)
        result = self.driver.get_screenshot_as_file(name)

        self.recover_screen()

        return result
``` 

What do you think?

Reply:
Thanks for the review!

The return value should be the filename otherwise user can't find it.
but, I still think this should be the user's responsibility to specify the path, so make the code easier to understand and predictable also more low coupling.

I mean it can be more readable if you write it something like this:

```py
browser.screenshot(get_temp_filename())

# helper
def get_temp_filename(prefix=name, suffix='.png'):
    (fd, name) = tempfile.mkstemp(prefix=name, suffix=suffix)
    os.close(fd)

    return name
```

Reply:
or maybe we can create the .png file in temp directory and then move it to the directory where user specified?

Reply:
I understand that. I'm just trying to think of something that wouldn't break backwards compatibility. In my idea you can still overwrite the file a call like the one you gave `browser.screenshot('~/latest.png')`  would work as you expect.

About the path: while the user should be responsible for a path, we can still provide a default temporary path and if the user really cares they can specify it. 

About what the function should return: why we should return the file path if the user is giving the file path to the function? Having a file descriptor, for example, makes it very easy to compare the screenshot with a test fixture. 

If we're breaking backwards compatibility, we could change from using `tempfile.mkstemp` to use `tempfile.NamedTemporaryFile` (https://docs.python.org/2/library/tempfile.html#tempfile.NamedTemporaryFile) and return a higher level file object.

Reply:
> In my idea you can still overwrite the file a call like the one you gave browser.screenshot('~/latest.png') would work as you expect.

I don't think this would work, the file name will change every time when calling the `tempfile.mkstemp`.

> About what the function should return: why we should return the file path if the user is giving the file path to the function?

I mean if you are going to use the `tempfile.mkstemp` then we should return the file path, because the file name changes every time. In this commit it return `True` if succeed or `False` if not(which is the value selenium returns).

Reply:
> I don't think this would work, the file name will change every time when calling the tempfile.mkstemp.

This function is only called whenever you don't give a `name` parameter to the function. If a name is not given, `tempfile.mkstemp` doesn't run. 

I see the problem then. And I think there is no solution on how to move forward without breaking backwards compatibility. The backwards compatibility is important though. Changing functions signature and return values needs to be done slowly. 

We should:

* Leave the current functions as they are today and add a deprecation warning, like done in #702. 
* Implement your proposed logic in new functions, with a different name.
* Release a new version of Splinter with a minor version increment with your added feature.
* When the deprecation has been there for long enough in the old functions, on the next major release (where breaking changes can be introduced) we drop them. 

Does this sound like a good plan for you, @yaquawa?

Reply:
Thanks! I'm all for that 👍🏼👍🏼

But I think we can just create another branch for major version, for example `0.11.x`, then keep the current code there, so we can just merge this into the `master`.
And then output the warning on branch `0.11.x`. (which I think it's the common way for handling such a case)

Reply:
Can you explain your plan better with real versions, I don't understand it well? For example, my plan is:

* Release 0.12.0 with your fix (will require that the functions are renamed to keep both behaviors), keep the old code and add a deprecation warning on it
* On the release 0.13.0, we just remove the old code.

This is exactly how phantomjs got removed. It started to output a deprecation warning in 0.8.0 and was removed in 0.9.0.

Reply:
For example like [Laravel](https://github.com/laravel/framework), they have the major versions for each branch.
So like what they do, we can create a branch called `0.11.x`; on that branch, this commit won't be merged at all, instead we add some warning to the `screenshot` method on that branch. For this commit, just merge it into the `master` branch.

Reply:
I'm sorry, but our model is not exactly the same as Laravel. Why not proceed as I proposed? I don't understand why you don't like my proposal... it's the way we used to work with breaking backwards compatibility and deprecations.

Reply:
I'm all ok for your proposal 👌
I just thought that the newly added method have to be renamed again may take a little effort too. Anyway I'm going to respect the way how you manage this repo!

Reply:
Thanks for the patience, @yaquawa. I'm sorry it was not clear until now. I'm eager to merge your contribution :)

Reply:
Thanks! @douglascamata
I've putted the old method back, and added the new one as `capture_screenshot`(which match the CDP's method name).